Stop Hazelcast client if server is unavailable

If the Hazelcast client loses its connection to the server then it retries to reconnect until the configured timeout is reached.

However, `ZeebeHazelcast` is still retrying to read from the ringbuffer. It ignores the exception that is thrown from the client.

```
16:06:06.917 [pool-2-thread-1] TRACE io.zeebe.hazelcast.connect.java.ZeebeHazelcast - Read from ring-buffer with sequence '0'
16:06:06.917 [pool-2-thread-1] ERROR io.zeebe.hazelcast.connect.java.ZeebeHazelcast - Fail to read from ring-buffer at sequence '0'. Will try again.
com.hazelcast.client.HazelcastClientNotActiveException: Client is shutting down
	at com.hazelcast.client.impl.spi.impl.ClientInvocation.notifyExceptionWithOwnedPermission(ClientInvocation.java:318) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.spi.impl.ClientInvocation.invokeOnSelection(ClientInvocation.java:211) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.spi.impl.ClientInvocation.invoke(ClientInvocation.java:158) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.proxy.ClientRingbufferProxy.invoke(ClientRingbufferProxy.java:219) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.proxy.ClientRingbufferProxy.readOne(ClientRingbufferProxy.java:156) ~[hazelcast-4.2.1.jar:4.2.1]
	at io.zeebe.hazelcast.connect.java.ZeebeHazelcast.readNext(ZeebeHazelcast.java:114) ~[classes/:?]
	at io.zeebe.hazelcast.connect.java.ZeebeHazelcast.readFromBuffer(ZeebeHazelcast.java:106) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
	at ------ submitted from ------.() ~[?:?]
	at com.hazelcast.internal.util.ExceptionUtil.cloneExceptionWithFixedAsyncStackTrace(ExceptionUtil.java:279) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.wrapRuntimeException(AbstractInvocationFuture.java:1968) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.wrapOrPeel(AbstractInvocationFuture.java:1949) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.spi.impl.AbstractInvocationFuture$ExceptionalResult.wrapForJoinInternal(AbstractInvocationFuture.java:1431) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.resolveAndThrowForJoinInternal(AbstractInvocationFuture.java:600) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.spi.impl.AbstractInvocationFuture.joinInternal(AbstractInvocationFuture.java:576) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.proxy.ClientRingbufferProxy.invoke(ClientRingbufferProxy.java:220) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.proxy.ClientRingbufferProxy.readOne(ClientRingbufferProxy.java:156) ~[hazelcast-4.2.1.jar:4.2.1]
	at io.zeebe.hazelcast.connect.java.ZeebeHazelcast.readNext(ZeebeHazelcast.java:114) ~[classes/:?]
	at io.zeebe.hazelcast.connect.java.ZeebeHazelcast.readFromBuffer(ZeebeHazelcast.java:106) ~[classes/:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.io.IOException: No connection found to cluster.
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.checkInvocationAllowed(TcpClientConnectionManager.java:542) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.spi.impl.ClientInvocationServiceImpl.checkInvocationAllowed(ClientInvocationServiceImpl.java:294) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.spi.impl.ClientInvocation.invokeOnSelection(ClientInvocation.java:180) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.spi.impl.ClientInvocation.invoke(ClientInvocation.java:158) ~[hazelcast-4.2.1.jar:4.2.1]
	at com.hazelcast.client.impl.proxy.ClientRingbufferProxy.invoke(ClientRingbufferProxy.java:219) ~[hazelcast-4.2.1.jar:4.2.1]
	... 8 more
```

--- a/connector-java/src/main/java/io/zeebe/hazelcast/connect/java/ZeebeHazelcast.java
+++ b/connector-java/src/main/java/io/zeebe/hazelcast/connect/java/ZeebeHazelcast.java
@@ -1,6 +1,7 @@
 package io.zeebe.hazelcast.connect.java;
 
 import com.google.protobuf.InvalidProtocolBufferException;
+import com.hazelcast.client.HazelcastClientNotActiveException;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.ringbuffer.Ringbuffer;
 import com.hazelcast.ringbuffer.StaleSequenceException;
@@ -76,6 +77,10 @@ public class ZeebeHazelcast implements AutoCloseable {
     future = executorService.submit(this::readFromBuffer);
   }
 
+  public boolean isClosed() {
+    return isClosed;
+  }
+
   /** Stop reading from the ringbuffer. */
   @Override
   public void close() throws Exception {
@@ -145,6 +150,15 @@ public class ZeebeHazelcast implements AutoCloseable {
           e);
 
       sequence = headSequence;
+
+    } catch (HazelcastClientNotActiveException e) {
+      LOGGER.warn("Lost connection to the Hazelcast server", e);
+
+      try {
+        close();
+      } catch (Exception closingFailure) {
+        LOGGER.debug("Failure while closing the client", closingFailure);
+      }
 
     } catch (InterruptedException e) {
       LOGGER.debug("Interrupted while reading from ring-buffer with sequence '{}'", sequence);

--- a/connector-java/src/test/java/io/zeebe/hazelcast/ZeebeHazelcastClientTest.java
+++ b/connector-java/src/test/java/io/zeebe/hazelcast/ZeebeHazelcastClientTest.java
@@ -1,0 +1,65 @@
+package io.zeebe.hazelcast;
+
+import com.hazelcast.client.HazelcastClient;
+import com.hazelcast.client.config.ClientConfig;
+import com.hazelcast.config.Config;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import io.zeebe.hazelcast.connect.java.ZeebeHazelcast;
+import io.zeebe.hazelcast.exporter.ExporterConfiguration;
+import org.awaitility.Awaitility;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.time.Duration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ZeebeHazelcastClientTest {
+
+  private static final ExporterConfiguration CONFIGURATION = new ExporterConfiguration();
+  private static final Duration CONNECTION_TIMEOUT = Duration.ofSeconds(5);
+
+  private ZeebeHazelcast zeebeHazelcast;
+
+  private HazelcastInstance hzInstance;
+  private HazelcastInstance hzClient;
+
+  @Before
+  public void init() {
+    final Config config = new Config();
+    config.getNetworkConfig().setPort(5702);
+    hzInstance = Hazelcast.newHazelcastInstance(config);
+
+    final ClientConfig clientConfig = new ClientConfig();
+
+    final var connectionRetryConfig =
+        clientConfig.getConnectionStrategyConfig().getConnectionRetryConfig();
+    connectionRetryConfig.setClusterConnectTimeoutMillis(CONNECTION_TIMEOUT.toMillis());
+
+    clientConfig.getNetworkConfig().addAddress("127.0.0.1:5702");
+    hzClient = HazelcastClient.newHazelcastClient(clientConfig);
+  }
+
+  @After
+  public void cleanUp() throws Exception {
+    zeebeHazelcast.close();
+    hzClient.shutdown();
+    hzInstance.shutdown();
+  }
+
+  @Test
+  public void shouldCloseIfHazelcastIsUnavailable() {
+    // given
+    zeebeHazelcast = ZeebeHazelcast.newBuilder(hzClient).build();
+
+    // when
+    hzInstance.shutdown();
+
+    // then
+    Awaitility.await()
+        .atMost(CONNECTION_TIMEOUT.multipliedBy(2))
+        .untilAsserted(() -> assertThat(zeebeHazelcast.isClosed()).isTrue());
+  }
+}

